testutils/infra/device.py

```diff
@@ -0,0 +1,217 @@
+    remote commands execution or sending/receiving files.
+    """
+
+    def __init__(self, host_string="localhost:8822", user="root"):
+        """Create a MenderDevice object-
+
+        Keyword arguments:
+        host_string -- Remote SSH host of the form host:port
+        user -- Remote SSH user
+        """
+        self.host, self.port = host_string.split(":")
+
+    @property
+    def host_string(self):
+        return "%s:%s" % (self.host, self.port)
+
+    def run(self, cmd, **kw):
+        """Run given cmd in remote SSH host
+
+        Argument:
+        cmd - sring with the command to execute remotely
+
+        Recognized keyword arguments:
+        hide - do not print stdout nor stderr, and do not fail on errors
+        warn_only - do not fail on errors
+        wait - timeout for how long to retry the execution
+
+        References for Fabric 2 migration:
+        - https://www.fabfile.org/upgrading.html#upgrading
+        - http://docs.pyinvoke.org/en/latest/api/runners.html#invoke.runners.Runner.run
+        - https://docs.fabfile.org/en/1.12.1/api/core/context_managers.html#fabric.context_managers.quiet
+        """
+        if kw.get("hide") == True:
+            del kw["hide"]
+            with fabric.api.quiet():
+                return self._run(cmd, **kw)
+        elif kw.get("warn_only") == True:
+            del kw["warn_only"]
+            with fabric.api.settings(warn_only=True):
+                return self._run(cmd, **kw)
+        else:
+            return self._run(cmd, **kw)
+
+    def _run(self, cmd, **kw):
+        if not fabric.api.env.host_string:
+            output = fabric.api.execute(self.run, cmd, hosts=self.host_string, **kw)
+            # Fabric's execute returns a dict with the output
+            # for each host. Return only our output
+            return output[self.host_string]
+
+        return _run(cmd, **kw)
+
+    def put(self, file, local_path=".", remote_path="."):
+        """Copy local_path/file into remote_path over SSH connection
+
+        Keyword arguments:
+        file - local filename
+        local_path - local dirpath
+        remote_path - remote dirpath
+        """
+        if not fabric.api.env.host_string:
+            fabric.api.execute(
+                self.put, file, local_path, remote_path, hosts=self.host_string
+            )
+            return
+
+        _put(file, local_path, remote_path)
+
+    def ssh_is_opened(self, wait=60 * 60):
+        """Block until SSH connection is established on the device
+
+        Keyword arguments:
+        wait - Timeout (in seconds)
+        """
+        self.run("true", hide=True, wait=wait)
+
+
+class MenderDeviceGroup:
+    """Group of SSH accessible devices with Mender client
+
+    Currently, run/put methods are serialized, but they will be migrated to a parallel execution with Fabric 2
+    https://docs.fabfile.org/en/latest/api/group.html
+    """
+
+    def __init__(self, host_string_list, user="root"):
```

Comment:
Question: 
Why can't the device group take a list of `MenderDevices` instead of the `host_string_list`?

Reply:
It is a bit coupled to the way DockerNamespace and classes gather the IP address of the clients. It is actually quite ugly (as the mentioned methods append a harcoded `:8822`) so I might do a fix on this but not in this PR ;)